### DISCUSSION TITLE
C# HUGE omission: no method call on zero references (like destructor)

### DISCUSSION THREAD

**Discussion by iperov** [dotnet/csharplang#9402] [General]
The era of immutable data has begun.

What does that mean? 
For example immutable object Image() has allocated memory. 
in C# If I call a transform chain like `.blur(...).contrast(...).resize(...)` it will create 3 copies of the modified images, and the allocated memory from `.blur()` will not be freed after `contrast()` on reference loss, but will accumulate, and will be freed maybe after an hour. 
This is not acceptable in 2k25.

The Image example is not the only one. Similarly, you can wrap GPU memory (tensors as in ML libraries), and perform operations. In GPU there is no way to wait even 1 second for memory to be freed, because memory is small and will quickly run out in hundreds of operations.

The code as `.blur(...).contrast(...).resize(...)` looks minimalistic and clear.
Even python calls `__del__()` on reference loss, allowing you to free memory on the fly without waiting for GC.

Currently in C# you can only handle this by explicitly calling `using` or `dispose`, which is bullshit because 3 one-line operations `.blur(...).contrast(...).resize(...)` would then take 9 lines.
This is why C# lost the ML race without even entering it because of this serious flaw.

You should have 5 years ago added the ability to call code when references are lost to zero, for example `override Object.OnReferenceLose()`.

**Comment by HaloFour** [comment]
I believe this is more a [runtime](https://github.com/dotnet/runtime) question.  The C# language specification does not specify what kind of garbage collection algorithm the runtime should provide, only that objects may be eventually deallocated when there are no more references.  That does not seem to preclude a reference counting algorithm that would free immediately.

**Reply by iperov** [reply]
what does runtime have to do with it? 
The language standard should guarantee code calls on reference loss, not different runtime implementations.

**Reply by HaloFour** [reply]
The runtime has everything to do with it as the language is separate from the environment on which it runs.  The only thing that the language specification states is that some kind of automatic garbage collection algorithm is provided by the runtime, and that there is a lot of latitude as to when that algorithm actually deallocates objects or calls finalizers.  Nothing precludes the runtime from providing a reference counting mechanism and deallocating immediately.  However, I expect that the amount of demand for it would be very low, and that there would be other concerns that would make it very impractical.

**Reply by iperov** [reply]
What kind of stupidity is that? 🤡
The concept of reference is part of the language, not runtime. Accordingly, reference has a counter. Which can logically become 0, i.e. nobody refers to the object anymore.
Overriding or not overriding a method is no more dangerous than unsafe code.

**Reply by HaloFour** [reply]
Incorrect, the concept of a reference is provided by the runtime, not the language.  How those references work, and how they are collected, are both a part of the runtime spec.  C# simply defers to the runtime to offer that behavior.

Also, I would suggest toning it down.  That kind of language is not appreciated.

**Reply by iperov** [reply]
you are trolling here or AI answered
https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/types/classes
*A type that is defined as a class is a reference type*

**Reply by HaloFour** [reply]
Yes it is, which extends `System.Object`, and that is provided by the runtime.  The language provides absolutely none of the internal implementation and is not involved in how garbage collection works.

**Reply by iperov** [reply]
Okay, go live in your world 🤡 stop flood here plz

**Reply by IS4Code** [reply]
You are both missing the real reason. C# cannot do this based on references because C# doesn't 100% know whether something is a reference or not:
```cs
int Test<T>() where T : new(), ISomething
{
    var obj = new T();
    return obj.Property;
}
```
Since `T` is a type parameter, it can be anything.

Now I think it is absolutely within the realm of possibility for the language to do something to address this:
* The compiler could provide hints to the runtime when a value gets out of scope. In some regards, especially in the `async` world, it might have better reasoning capabilities than the runtime.
* The compiler could offer a way to warn let you know in certain cases when a reference could be collected. This could utilize `IDisposable` neatly.
* The compiler could offer a syntax to use `IDisposable` in a way that composes better in expressions such as these. For example `using(image).using blur(...).using contrast(...).using resize(...)`. Probably not the best syntax but goes in the general direction.

**Reply by 333fred** [reply]
@iperov this style of commenting is not welcome here. Please familiarize yourself with our code of conduct: https://dotnetfoundation.org/about/policies/code-of-conduct.

**Reply by iperov** [reply]
@IS4Code
> using(image).using blur(...).using contrast(...).using resize(...)

yeah looks like a crutch 🩼

Also consider the case when we create a Tensor , and it can be dynamically used in one or more functions, i.e. we don't know in which ones, we just pass the tensor to the model, and the model depending on some factors may or may not use it, or pass it to another model, so using() using() won't work. At the same time, the memory should be freed as fast as possible. This is how PyTorch (ML library for python) works

We need to call the code on zero reference, and in no other way.
This feature was needed ASAP 5 years ago.

**Reply by HaloFour** [reply]
Python has a reference counting garbage collector, which is the only reason Python knows the number of references that any given instance has.  Of course that's not enough as reference counting can't handle cyclical references which is why Python also has a mark and sweep garbage collector.

.NET/C# only has a mark and sweep garbage collector in its runtime, so it has no idea how many references an instance might have.  It only cares about whether or not an instance can be traced back to a root reference, and any instance that can't is eligible for collection.  There's nothing that tracks the number of references to any given instance, nor is there anything that increments/decrements reference counts that could trigger an immediate deletion.  There isn't even a mechanism to request a manual deletion of an instance.

At best, from strictly the language perspective, you can try to work around those edges by providing a reference counting wrapper that can call `IDisposable.Dispose`.  But that would entail manually incrementing and decrementing any reference count.  Even if the language did try to automatically increment and decrement that wrapper for you, once you pass it into a separate assembly you lose all ability to track how it is used or referenced.

**Reply by huoyaoyuan** [reply]
It's totally impossible for a C# caller to determine the reference count to an object is zero. A method can certainly do `return new Self { RecurseReference = this; }` to keep a back reference, and **a caller has no access to callee's body**.

There are multiple threads discussing similar problem, for example https://github.com/dotnet/runtime/issues/92184. I remember there's a thread specifically about external resource like tensors but I can't find it now.

For pure managed object, it's certainly the runtime's responsibility to do optimizations, for example using [escape analysis](https://github.com/dotnet/runtime/pull/115388) to cancel allocation of temporary objects.

**Reply by iperov** [reply]
@huoyaoyuan What do you mean impossible? 
How does garbage collector determine that an object needs to call ~dtor() and then clear it from memory? There is a counter somewhere, so we should just call ~dtor() immediately at the place where the counter became zero.

**Reply by HaloFour** [reply]
> There is a counter somewhere

There are no counters in a [mark and sweep garbage collector](https://en.wikipedia.org/wiki/Tracing_garbage_collection).  The garbage collector goes through a "mark" phase where it traces all references from a root set of references.  Any instance that can be traced directly back to those root references are considered to be "alive", and those that aren't are eligible for collection.

**Reply by iperov** [reply]
@HaloFour holy... I didn't realize C# had such a terrible system. It turns out the language is dead :(

**Reply by IS4Code** [reply]
That's what virtually all modern programming languages use.

Also, as has been pointed out, C# does not care about how garbage collection works. This is on .NET.

**Reply by iperov** [reply]
I even just allocate heaps of memory in a loop, and at some point everything went into tens of gigabytes and won't free up.

```csharp
class Some11
{
    byte[] x;

    public Some11()
    {
        x = new byte[16 * 1024 * 1024];
    }

    ~Some11()
    {
        Console.WriteLine("asd");
    }
}

static class Program
{

    static void asd()
    {
        Some11? x = null;
        while (true)
        {
            Console.WriteLine("1");
            x = new Some11();
        }
```
How is this even possible in 2k25?
I don't understand what could be more logical than calling the destructor as soon as no one else owns the reference? Why did they invent this nonsense mark and sweep?

**Reply by huoyaoyuan** [reply]
Finalizer has its own queue to be invoked. The allocated array can only be collected *at the next GC* after the destructable object get finalized. As each objects holds 16 MB of memory, 10GB is just less than 1000 objects. It's very common to have thousands of objects pending for collection in GC-based system. Moreover, array of 16MB is allocated on the large object heap, which is collected less frequently.

**Finalizer is not destructor**. If you remove the finalizer, the memory consumption will be consistently less than 100MB.

**Reply by yaakov-h** [reply]
That's also using the workstation GC by default, I believe. Server GC should drop the memory consumption a bit more aggressively.

**Reply by huoyaoyuan** [reply]
To compare in ways how people would write program:

```cpp
class Holder
{
	char* ptr;

public:
	Holder() : ptr(nullptr) {}

	explicit Holder(size_t s)
	{
		ptr = new char[s];
	}

	~Holder()
	{
		delete[] ptr;
	}
};

void Test()
{
	std::vector<std::shared_ptr<Holder>> alive;
	alive.resize(1000);

	LARGE_INTEGER start, end;
	QueryPerformanceCounter(&start);
	for (int i = 0; i < 1000000; i++)
	{
		alive[i % 1000] = std::make_shared<Holder>(1000 + (i % 7) * 10);
	}
	QueryPerformanceCounter(&end);
	std::cout << end.QuadPart - start.QuadPart << " Ticks\n";
}

int main()
{
	while (true)
	{
		Test();
	}
}
```

```csharp
void Test()
{
    Holder[] alive = new Holder[1000];

    long start = Stopwatch.GetTimestamp();

    for (int i = 0; i < 1000000; i++)
    {
        alive[i % 1000] = new Holder(1000 + (i % 7) * 10);
    }

    long end = Stopwatch.GetTimestamp();

    Console.WriteLine($"{end - start} Ticks");

}

while (true)
{
    Test();
}

public class Holder(int size)
{
    byte[] array = new byte[size];
}
```

The unoptimized C# program (~420000 Ticks) simply outperforms the reference counting-based C++ program (~550000 Ticks). GC is very performant for how real-world application allocates memory: various sizes of small objects, some randomly lives longer.

**Comment by CyrusNajmabadi** [comment]
> You should have 5 years ago added the ability to call code when references are lost to zero, for example override Object.OnReferenceLose().

This does not appear to be a C# issue afaict.  You could consider opening a discussion with teh runtime.  They are the team that would have some concept of if an instance is 'referenced' and if/when those references are 'lost'.

**Comment by ds5678** [comment]
If these are common operations for you, you should just write some extension methods.

```cs
public static Image BlurAndDisposeCaller(this Image caller)
{
    var result = caller.blur();
    caller.Dispose();
    return result;
}
```

**Reply by iperov** [reply]
are you serious? 🤣

**Reply by yaakov-h** [reply]
You could also write the library in such a way that you chain the **operations** and then apply them all at once, or in a series of passes over the same image data. There's no hard requirement in computer science for each mutation of an image to produce a new copy of the image in the first place.

e.g. 
```cs
var newImage = image.Transform([Blur(...), Contrast(...), Resize(...)]);
// or
var newImage = image.Blur(...).Contrast(...).Resize(...).Apply();
```

**Reply by iperov** [reply]
yeah it's called graph based transformation. Tensorflow used the same approach but did not gain popularity. Now all research is written in pytorch. Tensorflow is dead.

**Reply by ds5678** [reply]
> are you serious? 🤣

It's a solution that can be used **today**. In your dream scenario where the language team decides to implement your idea (which won't happen because it's a runtime request), you would be waiting a minimum of 18 months because it's too late for it to be made part of C# 14.

**Reply by iperov** [reply]
>  team decides to implement your idea (which won't happen because it's a runtime request)

maybe that's a good thing. Only the best languages should survive.

**Comment by CyrusNajmabadi** [comment]
As mentioned: 

1. GC versus referencing counting are not decisions within the language. They are a runtime decision. So this is not an appropriate venue for this discussion. 

2. GC often excels *over* reference counting. So you'd need to make a pretty compelling case, with data, demonstrating the problem, and motivating ref counting as a solution. 

Regardless, this is not something the language cares about *unless* the runtime wants to do something special and believes there should be some special syntax to work with the ref counting system. 

My recommendation is that if you want to continue this, take it to the dotnet/runtime repo.
